### PR TITLE
Add confirmation dialog when performing dangerous actions in tournament client

### PR DESCRIPTION
**Pull request by ILW8**
Many "dangerous" actions in the tournament client do not require confirmation at all, and the buttons often are located near other oft-used buttons in the various editor screens. 

For example: 
- `Reset teams` button in the bracket editor screen's context menu
- `Clear all` button in the control panel of both Rounds editor and Team editor screens
- `Delete Round` button in Rounds editor screen
- `Delete Team` button in Team editor screen

This PR implements confirmation dialog for these dangerous actions and is mostly done by "copying" some of the dialog overlay code from OsuGame into TournamentGame.

This also adds tests to test the new dialogs. I couldn't find an existing example of a similar functionality nor test in the tournament client tests so I kinda winged it, hoping that it still meets the project's code quality standards.

- Feature
  - [x] `Reset teams` button in the bracket editor screen's context menu
  - [x] `Clear all` button in the control panel of both Rounds editor and Team editor screens
  - [x] `Delete Round` button in Rounds editor screen
  - [x] `Delete Team` button in Team editor screen
  - [x] add dialog overlay to editor screen test scenes

- Tests:
  - [x] test `Reset teams` button in the bracket editor screen's context menu

### CODE REVIEW

**Comment by bdach** [comment]
When it comes to testing tournament client there's probably no need to overdo it and test it zealously. If the changes you make to add the confirmation dialogs are obvious (which they should be), we should be fine without testing.

File: osu.Game.Tournament.Tests/Screens/TestSceneLadderEditorScreen.cs [hunk 1]
```diff
@@ -1,22 +1,99 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
-using osu.Framework.Allocation;
+#nullable disable
```
Comment (bdach) [inline osu.Game.Tournament.Tests/Screens/TestSceneLadderEditorScreen.cs:4]: Please don't use `#nullable disable` in new code unless there's a very good reason to do so.

File: osu.Game.Tournament/Screens/Editors/Components/LadderResetTeamsDialog.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Overlays.Dialog;
+using osu.Game.Tournament.Screens.Ladder.Components;
+
+namespace osu.Game.Tournament.Screens.Editors.Components
+{
+    public partial class LadderResetTeamsDialog : DangerousActionDialog
+    {
+        public LadderResetTeamsDialog(Container<DrawableTournamentMatch> matchesContainer)
```
Comment (bdach) [inline osu.Game.Tournament/Screens/Editors/Components/LadderResetTeamsDialog.cs:13]: Please don't pass containers to these dialogs like this, pass the `Action` to execute instead.

Hell I'm not sure you even need to make all these classes. You could probably use `DangerousActionDialog` on its own and locally set `HeaderText`, `Icon` and `DangerousAction` as you see fit.

File: osu.Game.Tournament/Screens/Editors/Components/LadderResetTeamsDialog.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        {
+            HeaderText = @"Confirm reset teams?";
```
Comment (bdach) [inline osu.Game.Tournament/Screens/Editors/Components/LadderResetTeamsDialog.cs:15]:
```suggestion
            HeaderText = @"Reset teams?";
```

I can't say I like this phrasing with "confirm" jammed in there at the start. Reads wrong.

File: osu.Game.Tournament/Screens/Editors/Components/TournamentClearAllDialog.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+using osu.Game.Overlays.Dialog;
+
+namespace osu.Game.Tournament.Screens.Editors.Components
+{
+    public partial class TournamentClearAllDialog : DangerousActionDialog
+
+    {
+        public TournamentClearAllDialog(IList storage)
+        {
+            HeaderText = @"Confirm clear all?";
```
Comment (bdach) [inline osu.Game.Tournament/Screens/Editors/Components/TournamentClearAllDialog.cs:15]:
```suggestion
            HeaderText = @"Clear all?";
```

Mind you this text is still wonderfully unspecific but making it specific is going to be somewhat painful due to surrounding structure of `TournamentEditorScreen` so probably fine to stop at the generic option...

File: osu.Game.Tournament/Screens/Editors/LadderEditorScreen.cs [hunk 5]
```diff
@@ -30,6 +33,10 @@ public partial class LadderEditorScreen : LadderScreen, IHasContextMenu
 
         private WarningBox rightClickMessage;
 
+        [Resolved]
+        [CanBeNull]
```
Comment (bdach) [inline osu.Game.Tournament/Screens/Editors/LadderEditorScreen.cs:37]: This - unfortunately - doesn't quite work like you'd like it too. Yes the `[CanBeNull]` technically means that resharper will null-check the property, _however_ `[Resolved]` without `CanBeNull = true` set inside means that DI will crash if it fails to resolve the dialog overlay. So in practice, `dialogOverlay` will either not be null, or the game will crash.

The proper definition is

```csharp
            [Resolved(canBeNull: true)]
            [CanBeNull]
            private IDialogOverlay dialogOverlay { get; set; }
```

With NRT this gets a lot less boilerplatey, but we can't go there yet in tournament client.

File: osu.Game.Tournament/Screens/Editors/TournamentEditorScreen.cs [hunk 6]
```diff
@@ -27,6 +30,10 @@ public abstract partial class TournamentEditorScreen<TDrawable, TModel> : Tourna
     {
         protected abstract BindableList<TModel> Storage { get; }
 
+        [Resolved]
+        [CanBeNull]
+        private IDialogOverlay dialogOverlay { get; set; }
```
Comment (bdach) [inline osu.Game.Tournament/Screens/Editors/TournamentEditorScreen.cs:35]: As above

File: osu.Game.Tournament/TournamentGame.cs [hunk 7]
```diff
@@ -90,12 +101,14 @@ private void load(FrameworkConfigManager frameworkConfig, GameHost host)
...
                         Child = new TournamentSceneManager()
-                    }
+                    },
+                    topMostOverlayContent = new Container { RelativeSizeAxes = Axes.Both }
                 }, drawables =>
                 {
                     loadingSpinner.Hide();
                     loadingSpinner.Expire();
 
+                    topMostOverlayContent.Add(dialogOverlay);
```
Comment (bdach) [inline osu.Game.Tournament/TournamentGame.cs:111]: This is weird, why is `dialogOverlay` synchronously added on the load complete callback of the _async load_? Can you put the `dialogOverlay` as a child of `topMostOverlayContent` from the start?

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game.Tournament/TournamentGame.cs (hunk 7 above)
Comment (ILW8) [reply osu.Game.Tournament/TournamentGame.cs:111]: This is the part I was kinda winging, I'm not sure how I ended up with the code in this form; IIRC doing what felt intuitive to me did not seem to work. I'll incorporate your other suggested changes first and revisit this soon

File: osu.Game.Tournament/Screens/Editors/Components/LadderResetTeamsDialog.cs (hunk 2 above)
Comment (ILW8) [reply osu.Game.Tournament/Screens/Editors/Components/LadderResetTeamsDialog.cs:13]: `DangerousActionDialog` is declared abstract and its constructor as well as DangerousAction are protected. It looks like `DangerousAction` wasn't intended to be used directly and was meant to be subclassed. 
Of course I could remove the `abstract` modifiers and change both ctor and DangerousAction to public, thoughts?

Comment (bdach) [reply osu.Game.Tournament/Screens/Editors/Components/LadderResetTeamsDialog.cs]: I guess I didn't notice _that_, somehow.

Probably keep the classes. Note about passing `Action` still applies, though.

**Review by bdach — DISMISSED** [review]

**Review by peppy — APPROVED** [review]

**Pull request merged**